parser/src/elements/atomic/paragraph.rs

```diff
@@ -54,21 +55,24 @@ impl From<Vec<&'_ Symbol<'_>>> for Paragraph {
     }
 }
 
-fn not_closing_symbol(symbol: &&Symbol) -> bool {
-    [SymbolKind::Blankline, SymbolKind::EOI]
-        .iter()
-        .all(|closing| *closing != symbol.kind)
-}
-
 impl ElementParser for Paragraph {
     type Token<'a> = &'a Symbol<'a>;
 
     fn tokenize<'i>(input: &mut SymbolIterator<'i>) -> Option<TokenizeOutput<Self::Token<'i>>> {
-        let content = input.by_ref().take_while(not_closing_symbol).collect();
-
-        let output = TokenizeOutput { tokens: content };
+        let mut content_iter = input.nest(
+            None,
+            Some(Rc::new(|matcher: &mut dyn EndMatcher| {
+                matcher.consumed_is_empty_line()
+            })),
+        );
+        let content = content_iter.take_to_end();
+        content_iter.update(input);
```

Comment:
It just occurred to me how could we (maybe) enforce update of `input` iterator. 

We could make it so that `take_to_end` function takes the `input` iterator as parameter. The function returns `Option<Vec<Symbol>>`: 

1. Consumes symbols until the end
2. Checks whether the end is reached
   * if end is reached
     - update `input` iterator
     - `return Some(symbols)`
   * if end is not reached 
     - do not update `input` iterator
     - `return None`

Then, to get the content, we have to provide the original input iterator. And we can then automatically update the input iterator, preventing the bug where we might forget to update the input iterator. What do you think? 🤔

Reply:
It must be able to call `take_to_end()` without passing an iterator, because it is possible to set end matching without nesting an iterator.

I would leave it as is right now.
Maybe once we implement more elements we see a better option on how to enforce the update.